arcgis-ios-sdk-samples/Route & Directions/Navigate route/NavigateRouteViewController.swift

```diff
@@ -0,0 +1,296 @@
+                self.setRouteTracker(result)
+                self.setRouteGraphics(result)
+                self.setSpeechSynthesizer()
+                // Enable bar button item.
+                self.navigateButtonItem.isEnabled = true
+                _ = routeTask
+            }
+        }
+    }
+    
+    /// Sets a speech synthesizer to generate voice guidance based on text.
+    func setSpeechSynthesizer() {
+        speechSynthesizer = AVSpeechSynthesizer()
+    }
+    
+    /// Sets the simulated data source for this demo.
+    /// - Parameter result: solved route from the route task.
+    func setDataSource(_ result: AGSRouteResult) {
+        // datasource should not be time-dependent
+        if let route = result.routes.first {
+            directionsList = route.directionManeuvers
+            let densifiedRoute = AGSGeometryEngine.geodeticDensifyGeometry(route.routeGeometry!, maxSegmentLength: 50.0, lengthUnit: .meters(), curveType: .geodesic) as! AGSPolyline
+            mockDataSource = AGSSimulatedLocationDataSource()
+            mockDataSource!.setLocationsWith(densifiedRoute)
+            mockDataSource!.locationChangeHandlerDelegate = self
+            mapView.locationDisplay.dataSource = mockDataSource!
+        }
+    }
+    
+    /// Sets the route tracker to provide navigation information.
+    /// - Parameter result: solved route from the route task.
+    func setRouteTracker(_ result: AGSRouteResult) {
+        routeTracker = AGSRouteTracker(routeResult: result, routeIndex: 0)
+        routeTracker!.delegate = self
+    }
+    
+    /// Sets the graphics.
+    /// - Parameter result: solved route from the route task.
+    func setRouteGraphics(_ result: AGSRouteResult) {
+        if let route = result.routes.first {
+            let stopSymbol = AGSSimpleMarkerSymbol(style: .diamond, color: .orange, size: 20)
+            for stop in makeStops() {
+                graphicsOverlay.graphics.add(AGSGraphic(geometry: stop.geometry, symbol: stopSymbol))
+            }
+            // Show the resulting route on the map and save a reference to the route.
+            routeAheadGraphic.geometry = route.routeGeometry
+            routeTraveledGraphic.geometry = nil
+            graphicsOverlay.graphics.addObjects(from: [
+                routeAheadGraphic,
+                routeTraveledGraphic
+            ])
+            if let routeGeometry = route.routeGeometry {
+                if let viewPoint = defaultViewPoint {
+                    // Reset to initial view point with animation.
+                    mapView.setViewpoint(viewPoint, completion: nil)
+                } else {
+                    mapView.setViewpointGeometry(routeGeometry) { [weak self] _ in
+                        // Get the initial zoomed view point.
+                        self?.defaultViewPoint = self?.mapView.currentViewpoint(with: .centerAndScale)
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Again, with a name like `setRouteGraphics(_:)`, I expected this method to take an array of graphics, just like `AGSRouteParameters.setStops(_:)` takes an array of stops. It also seems like it is trying to do too many things. Here are a few thoughts:

- One of the things this method does is create graphics and add them to the graphics overlay. When the route is reset, all the graphics are removed and then this method is called to recreate them. Unless I'm missing something, all of the stop graphics are always the same, so they don't need to be recreated every time. And `routeAheadGraphic` and `routeTraveledGraphic` are always the same object, just their geometries are updated. That got me thinking: could all of this be extracted into a method that creates the graphics, creates a graphics overlay and adds the graphics to it, and then returns the graphics overlay (`func makeRouteOverlay() -> AGSGraphicsOverlay`? Then, when the route is solved, the graphics overlay could be created and added to the map view, eliminating the need for the `graphicsOverlay` property.

- Another thing this method does is update the geometry of `routeAheadGraphic` and `routeTraveledGraphic`. Perhaps that could be extracted into a method that updates the route graphics? It could take a tracking progress: `func updateRouteGraphics(trackingProgress: AGSTrackingProgress)`.

- The third thing it does is zoom update the viewpoint. Perhaps that should be a third method.